cpp/src/arrow/util/thread_pool.h

```diff
@@ -102,16 +102,34 @@ class ARROW_EXPORT Executor {
   // CPU heavy work off the I/O thread pool.  So the I/O task should transfer
   // the future to the CPU executor before returning.
   template <typename T>
-  Future<T> Transfer(Future<T> future) {
+  Future<T> Transfer(Future<T> future, bool force_spawn = false) {
     auto transferred = Future<T>::Make();
-    future.AddCallback([this, transferred](const Result<T>& result) mutable {
+    auto callback = [this, transferred](const Result<T>& result) mutable {
       auto spawn_status = Spawn([transferred, result]() mutable {
         transferred.MarkFinished(std::move(result));
       });
       if (!spawn_status.ok()) {
         transferred.MarkFinished(spawn_status);
       }
-    });
+    };
+    auto callback_factory = [&callback]() { return callback; };
+    auto callback_added = future.TryAddCallback(callback_factory);
+    if (!callback_added) {
+      if (force_spawn) {
+        auto spawn_status = Spawn([future, transferred]() mutable {
+          transferred.MarkFinished(future.result());
+        });
+        if (!spawn_status.ok()) {
+          transferred.MarkFinished(spawn_status);
+        }
+        return transferred;
```

Comment:
I'm a little confused what the new parameter accomplishes. Since the new future essentially completes immediately anyways, there's still no guarantee that future callers of transferred.AddCallback won't just have their callbacks synchronously executed as soon as they add them on the current thread.

Reply:
Haha, it made sense in my head but you're right, this isn't the right approach.  I'm confused as to how it worked but I guess it helped inject a little bit of opportunity in the system because it would take a little while for the spwaned thread to mark the callback finished so it slowed it down a little.  Ok, I'm going to try something different.

Reply:
Ok.  The "force_spawn" has been removed.  Using the instrumentation I worked on yesterday I tracked the issue down to a few other spots that didn't really have anything to do with needing to transfer.  On the bright side, this actually eliminated the 2-5% slowdown I was seeing so the performance now matches the synchronous version.